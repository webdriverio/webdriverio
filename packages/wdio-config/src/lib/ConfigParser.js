import fs from 'fs'
import path from 'path'
import glob from 'glob'
import merge from 'deepmerge'

import logger from 'wdio-logger'

import { detectBackend } from '../utils'

import { DEFAULT_CONFIGS, SUPPORTED_HOOKS } from '../constants'

const log = logger('wdio-config:ConfigParser')
const MERGE_OPTIONS = { clone: false }

export default class ConfigParser {
    constructor () {
        this._config = DEFAULT_CONFIGS
        this._capabilities = []
    }

    /**
     * merges config file with default values
     * @param {String} filename path of file relative to current directory
     */
    addConfigFile (filename) {
        if (typeof filename !== 'string') {
            throw new Error('addConfigFile requires filepath')
        }

        var filePath = path.resolve(process.cwd(), filename)

        try {
            /**
             * clone the original config
             */
            var fileConfig = merge(require(filePath).config, {}, MERGE_OPTIONS)

            /**
             * merge capabilities
             */
            const defaultTo = Array.isArray(this._capabilities) ? [] : {}
            this._capabilities = merge(this._capabilities, fileConfig.capabilities || defaultTo, MERGE_OPTIONS)
            delete fileConfig.capabilities

            /**
             * add service hooks and remove them from config
             */
            this.addService(fileConfig)
            for (let hookName of SUPPORTED_HOOKS) {
                delete fileConfig[hookName]
            }

            this._config = merge(this._config, fileConfig, MERGE_OPTIONS)

            /**
             * detect Selenium backend
             */
            this._config = merge(detectBackend(this._config), this._config, MERGE_OPTIONS)
        } catch (e) {
            log.error(`Failed loading configuration file: ${filePath}`)
            throw e
        }
    }

    /**
     * merge external object with config object
     * @param  {Object} object  desired object to merge into the config object
     */
    merge (object = {}) {
        this._config = merge(this._config, object, MERGE_OPTIONS)
        let spec = Array.isArray(object.spec) ? object.spec : []
        let exclude = Array.isArray(object.exclude) ? object.exclude : []

        /**
         * overwrite config specs that got piped into the wdio command
         */
        if (object.specs && object.specs.length > 0) {
            this._config.specs = object.specs
        } else if (object.exclude && object.exclude.length > 0) {
            this._config.exclude = object.exclude
        }

        /**
         * merge capabilities
         */
        const defaultTo = Array.isArray(this._capabilities) ? [] : {}
        this._capabilities = merge(this._capabilities, this._config.capabilities || defaultTo, MERGE_OPTIONS)

        /**
         * run single spec file only, regardless of multiple-spec specification
         */
        this.setFilePathToFilterOptions(spec)
        this.setFilePathToFilterOptions(exclude, true)

        /**
         * user and key could get added via cli arguments so we need to detect again
         * Note: cli arguments are on the right and overwrite config
         * if host and port are default, remove them to get new values
         */
        let defaultBackend = detectBackend({})
        if (this._config.hostname === defaultBackend.hostname && this._config.port === defaultBackend.port) {
            delete this._config.hostname
            delete this._config.port
        }

        this._config = merge(detectBackend(this._config), this._config, MERGE_OPTIONS)
    }

    /**
     * add hooks from services to runner config
     * @param {Object} service  a service is basically an object that contains hook methods
     */
    addService (service) {
        for (let hookName of SUPPORTED_HOOKS) {
            if (!service[hookName]) {
                continue
            }

            if (typeof service[hookName] === 'function') {
                this._config[hookName].push(service[hookName].bind(service))
            } else if (Array.isArray(service[hookName])) {
                for (let hook of service[hookName]) {
                    if (typeof hook === 'function') {
                        this._config[hookName].push(hook.bind(service))
                    }
                }
            }
        }
    }

    /**
     * get excluded files from config pattern
     */
    getSpecs (capSpecs, capExclude) {
        let specs = ConfigParser.getFilePaths(this._config.specs)
        let spec  = Array.isArray(this._config.spec) ? this._config.spec : []
        let exclude = ConfigParser.getFilePaths(this._config.exclude)
        let suites = Array.isArray(this._config.suite) ? this._config.suite : []

        /**
         * check if user has specified a specific suites to run
         */
        if (suites.length > 0) {
            let suiteSpecs = []
            for (let suiteName of suites) {
                // ToDo: log warning if suite was not found
                let suite = this._config.suites[suiteName]
                if (suite && Array.isArray(suite)) {
                    suiteSpecs = suiteSpecs.concat(ConfigParser.getFilePaths(suite))
                }
            }

            if (suiteSpecs.length === 0) {
                throw new Error(`The suite(s) "${suites.join('", "')}" you specified don't exist ` +
                                'in your config file or doesn\'t contain any files!')
            }

            // Allow --suite and --spec to both be defined on the command line
            // Removing any duplicate tests that could be included
            const tmp_specs = spec.length > 0 ? [...specs, ...suiteSpecs] : suiteSpecs

            return [...new Set(tmp_specs)]
        }

        if (Array.isArray(capSpecs)) {
            specs = specs.concat(ConfigParser.getFilePaths(capSpecs))
        }

        if (Array.isArray(capExclude)) {
            exclude = exclude.concat(ConfigParser.getFilePaths(capExclude))
        }

        return specs.filter(spec => exclude.indexOf(spec) < 0)
    }

    /**
     * sets config attribute with file paths from filtering
     * options from cli argument
     *
     * @param  {String} filterObj  list of files in a string from
     * cli argument
     * @return {Boolean} exclude  true or false
     */
    setFilePathToFilterOptions (filterObj, exclude = false) {
        const FILTER_TYPE = exclude ? 'exclude' : 'specs'
        if (filterObj.length > 0) {
            const objs = new Set()
            const objsList = ConfigParser.getFilePaths(this._config[FILTER_TYPE])
            filterObj.forEach(filtered_file => {
                if (fs.existsSync(filtered_file) && fs.lstatSync(filtered_file).isFile()) {
                    objs.add(path.resolve(process.cwd(), filtered_file))
                }
                else {
                    objsList.forEach(file => {
                        if (file.match(filtered_file)) {
                            objs.add(file)
                        }
                    })
                }
            })
            if (objs.size === 0) {
                throw new Error(`spec file(s) ${filterObj.join(`, `)} not found`)
            }
            this._config[FILTER_TYPE] = [...objs]
        }
    }

    /**
     * return configs
     */
    getConfig () {
        return this._config
    }

    /**
     * return capabilities
     */
    getCapabilities (i) {
        if (typeof i === 'number' && this._capabilities[i]) {
            return this._capabilities[i]
        }

        return this._capabilities
    }

    /**
     * returns a flatten list of globed files
     *
     * @param  {String[]} filenames  list of files to glob
     * @return {String[]} list of files
     */
    static getFilePaths (patterns, omitWarnings) {
        let files = []

        if (typeof patterns === 'string') {
            patterns = [patterns]
        }

        if (!Array.isArray(patterns)) {
            throw new Error('specs or exclude property should be an array of strings')
        }

        for (let pattern of patterns) {
            let filenames = glob.sync(pattern)

            filenames = filenames.filter(filename =>
                filename.slice(-3) === '.js' ||
                filename.slice(-3) === '.ts' ||
                filename.slice(-8) === '.feature' ||
                filename.slice(-7) === '.coffee')

            filenames = filenames.map(filename =>
                path.isAbsolute(filename) ? path.normalize(filename) : path.resolve(process.cwd(), filename))

            if (filenames.length === 0 && !omitWarnings) {
                log.warn('pattern', pattern, 'did not match any file')
            }

            files = merge(files, filenames, MERGE_OPTIONS)
        }

        return files
    }
}
