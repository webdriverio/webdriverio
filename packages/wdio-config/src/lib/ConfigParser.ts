import fs from 'fs'
import path from 'path'
import glob from 'glob'
import merge from 'deepmerge'
import logger from '@wdio/logger'

import { detectBackend, removeLineNumbers, isCucumberFeatureWithLineNumber, validObjectOrArray } from '../utils'
import { DEFAULT_CONFIGS, SUPPORTED_HOOKS, SUPPORTED_FILE_EXTENSIONS } from '../constants'
import type { Capabilities, ConfigOptions, Hooks } from '../types'

const log = logger('@wdio/config:ConfigParser')
const MERGE_OPTIONS = { clone: false }

export default class ConfigParser {
    private _config: ConfigOptions = DEFAULT_CONFIGS()
    private _capabilities: Capabilities = [];

    /**
     * merges config file with default values
     * @param {String} filename path of file relative to current directory
     */
    addConfigFile (filename: string) {
        if (typeof filename !== 'string') {
            throw new Error('addConfigFile requires filepath')
        }

        const filePath = path.resolve(process.cwd(), filename)

        try {
            /**
             * clone the original config
             */
            const fileConfig = merge<ConfigOptions>(require(filePath).config, {}, MERGE_OPTIONS)

            /**
             * merge capabilities
             */
            const defaultTo: Capabilities = Array.isArray(this._capabilities) ? [] : {}
            this._capabilities = merge<Capabilities>(this._capabilities, fileConfig.capabilities || defaultTo, MERGE_OPTIONS)
            delete fileConfig.capabilities

            /**
             * Add hooks from the file config and remove them from file config object to avoid
             * complications when using merge function
             */
            this.addService(fileConfig)
            for (let hookName of SUPPORTED_HOOKS) {
                delete fileConfig[hookName]
            }

            this._config = merge(this._config, fileConfig, MERGE_OPTIONS)

            /**
             * For Sauce Labs RDC we need to determine if the config file has a `testobject_api_key`
             * If so, we need to provide a boolean to the `detectBackend` to set the correct hostname
             *
             * NOTE: This will not work for multi remote
             */
            const isRDC = Array.isArray(this._capabilities) && this._capabilities.some(capability => 'testobject_api_key' in capability)

            /**
             * detect Selenium backend
             */
            this._config = merge(detectBackend(this._config, isRDC), this._config, MERGE_OPTIONS)

            /**
             * remove `watch` from config as far as it can be only passed as command line argument
             */
            delete this._config.watch
        } catch (e) {
            log.error(`Failed loading configuration file: ${filePath}:`, e.message)
            throw e
        }
    }

    /**
     * merge external object with config object
     * @param  {Object} object  desired object to merge into the config object
     */
    merge (object: Partial<ConfigOptions> = {}) {
        const spec = Array.isArray(object.spec) ? object.spec : []
        const exclude = Array.isArray(object.exclude) ? object.exclude : []

        this._config = merge(this._config, object, MERGE_OPTIONS)
        /**
         * overwrite config specs that got piped into the wdio command
         */
        if (object.specs && object.specs.length > 0) {
            this._config.specs = object.specs
        } else if (object.exclude && object.exclude.length > 0) {
            this._config.exclude = object.exclude
        }

        /**
         * overwrite capabilities
         */
        this._capabilities = validObjectOrArray(this._config.capabilities) ? this._config.capabilities : this._capabilities
        /**
         * save original specs if Cucumber's feature line number is provided
         */
        if (this._config.spec && isCucumberFeatureWithLineNumber(this._config.spec)) {
            /**
             * `this._config.spec` is string instead of Array in watch mode
             */
            this._config.cucumberFeaturesWithLineNumbers = Array.isArray(this._config.spec) ? [...this._config.spec] : [this._config.spec]
        }

        /**
         * run single spec file only, regardless of multiple-spec specification
         */
        if (spec.length > 0) {
            this._config.specs = this.setFilePathToFilterOptions(spec, this._config.specs!)
        }
        if (exclude.length > 0) {
            this._config.exclude = this.setFilePathToFilterOptions(exclude, this._config.exclude!)
        }

        this._config = merge(detectBackend(this._config), this._config, MERGE_OPTIONS)
    }

    /**
     * Add hooks from an existing service to the runner config.
     * @param {Object} service - an object that contains hook methods.
     */
    addService (service: Hooks) {
        const addHook = <T extends keyof Hooks>(hookName: T, hook: Extract<Hooks[T], Function>) => {
            const existingHooks: ConfigOptions[keyof Hooks] = this._config[hookName]
            if (!existingHooks) {
                // @ts-ignore Expression produces a union type that is too complex to represent
                this._config[hookName] = hook.bind(service)
            } else if (typeof existingHooks === 'function') {
                this._config[hookName] = [existingHooks, hook.bind(service)]
            } else {
                this._config[hookName] = [...existingHooks, hook.bind(service)]
            }
        }

        for (const hookName of SUPPORTED_HOOKS) {
            const hooksToBeAdded = service[hookName]
            if (!hooksToBeAdded) {
                continue
            }

            if (typeof hooksToBeAdded === 'function') {
                addHook(hookName, hooksToBeAdded)
            } else if (Array.isArray(hooksToBeAdded)) {
                for (const hookToAdd of hooksToBeAdded) {
                    if (typeof hookToAdd === 'function') {
                        addHook(hookName, hookToAdd)
                    }
                }
            }
        }
    }

    /**
     * get excluded files from config pattern
     */
    getSpecs (capSpecs?: string[], capExclude?: string[]) {
        let specs = ConfigParser.getFilePaths(this._config.specs!)
        let spec  = Array.isArray(this._config.spec) ? this._config.spec : []
        let exclude = ConfigParser.getFilePaths(this._config.exclude!)
        let suites = Array.isArray(this._config.suite) ? this._config.suite : []

        /**
         * check if user has specified a specific suites to run
         */
        if (suites.length > 0) {
            let suiteSpecs: string[] = []
            for (let suiteName of suites) {
                let suite = this._config.suites?.[suiteName]
                if (!suite) {
                    log.warn(`No suite was found with name "${suiteName}"`)
                }
                if (Array.isArray(suite)) {
                    suiteSpecs = suiteSpecs.concat(ConfigParser.getFilePaths(suite))
                }
            }

            if (suiteSpecs.length === 0) {
                throw new Error(`The suite(s) "${suites.join('", "')}" you specified don't exist ` +
                                'in your config file or doesn\'t contain any files!')
            }

            // Allow --suite and --spec to both be defined on the command line
            // Removing any duplicate tests that could be included
            let tmpSpecs = spec.length > 0 ? [...specs, ...suiteSpecs] : suiteSpecs

            if (Array.isArray(capSpecs)) {
                tmpSpecs = ConfigParser.getFilePaths(capSpecs)
            }

            if (Array.isArray(capExclude)) {
                exclude = ConfigParser.getFilePaths(capExclude)
            }

            specs = [...new Set(tmpSpecs)]
            return specs.filter(spec => !exclude.includes(spec))
        }

        if (Array.isArray(capSpecs)) {
            specs = ConfigParser.getFilePaths(capSpecs)
        }

        if (Array.isArray(capExclude)) {
            exclude = ConfigParser.getFilePaths(capExclude)
        }

        return specs.filter(spec => !exclude.includes(spec))
    }

    /**
     * sets config attribute with file paths from filtering
     * options from cli argument
     *
     * @param  {String} cliArgFileList  list of files in a string from
     * @param  {Object} config  config object that stores the spec and exclude attributes
     * cli argument
     * @return {String[]} List of files that should be included or excluded
     */
    setFilePathToFilterOptions (cliArgFileList: string[], config: string[]) {
        const filesToFilter = new Set<string>()
        const fileList = ConfigParser.getFilePaths(config)
        cliArgFileList.forEach(filteredFile => {
            filteredFile = removeLineNumbers(filteredFile)
            let globMatchedFiles = ConfigParser.getFilePaths(glob.sync(filteredFile))
            if (fs.existsSync(filteredFile) && fs.lstatSync(filteredFile).isFile()) {
                filesToFilter.add(path.resolve(process.cwd(), filteredFile))
            } else if (globMatchedFiles.length) {
                globMatchedFiles.forEach(file => filesToFilter.add(file))
            } else {
                fileList.forEach(file => {
                    if (file.match(filteredFile)) {
                        filesToFilter.add(file)
                    }
                })
            }
        })
        if (filesToFilter.size === 0) {
            throw new Error(`spec file(s) ${cliArgFileList.join(', ')} not found`)
        }
        return [...filesToFilter]
    }

    /**
     * return configs
     */
    getConfig () {
        return this._config as Required<ConfigOptions>
    }

    /**
     * return capabilities
     */
    getCapabilities (i?: number) {
        if (typeof i === 'number' && Array.isArray(this._capabilities) && this._capabilities[i]) {
            return this._capabilities[i]
        }

        return this._capabilities
    }

    /**
     * returns a flatten list of globed files
     *
     * @param  {String[]} filenames  list of files to glob
     * @return {String[]} list of files
     */
    static getFilePaths (patterns: string[], omitWarnings?: boolean) {
        let files: string[] = []

        if (typeof patterns === 'string') {
            patterns = [patterns]
        }

        if (!Array.isArray(patterns)) {
            throw new Error('specs or exclude property should be an array of strings')
        }

        patterns = patterns.map(pattern => removeLineNumbers(pattern))

        for (let pattern of patterns) {
            let filenames = glob.sync(pattern)

            filenames = filenames.filter(
                (filename) => SUPPORTED_FILE_EXTENSIONS.find(
                    (ext) => filename.endsWith(ext)))

            filenames = filenames.map(filename =>
                path.isAbsolute(filename) ? path.normalize(filename) : path.resolve(process.cwd(), filename))

            if (filenames.length === 0 && !omitWarnings) {
                log.warn('pattern', pattern, 'did not match any file')
            }

            files = merge(files, filenames, MERGE_OPTIONS)
        }

        return files
    }
}
