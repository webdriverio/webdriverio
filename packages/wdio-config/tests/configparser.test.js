import path from 'path'
import ConfigParser from '../src/lib/ConfigParser'

const FIXTURES_PATH = path.resolve(__dirname, '__fixtures__')
const FIXTURES_CONF = path.resolve(FIXTURES_PATH, 'wdio.conf.js')
const FIXTURES_CONF_RDC = path.resolve(FIXTURES_PATH, 'wdio.conf.rdc.js')
const FIXTURES_CONF_MULTIREMOTE_RDC = path.resolve(FIXTURES_PATH, 'wdio.conf.multiremote.rdc.js')
const FIXTURES_LOCAL_CONF = path.resolve(FIXTURES_PATH, 'wdio.local.conf.js')
const INDEX_PATH = path.resolve(__dirname, '..', 'src', 'index.js')

describe('ConfigParser', () => {
    it('should throw if getFilePaths is not a string', () => {
        expect(() => ConfigParser.getFilePaths(123)).toThrow()
    })

    describe('addConfigFile', () => {
        it('should throw if config file is not a string', () => {
            const configParser = new ConfigParser()
            expect(() => configParser.addConfigFile(123)).toThrow()
        })

        it('should throw if config file does not exist', () => {
            const configParser = new ConfigParser()
            expect(() => configParser.addConfigFile(path.resolve(__dirname, 'foobar.conf.js'))).toThrow()
        })

        it('should add the rdc hostname when a rdc conf is provided', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF_RDC)
            expect(configParser._config.hostname).toContain('appium.testobject.com')
        })

        it('should default to the vm hostname when a multiremote conf with rdc props is provided', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF_MULTIREMOTE_RDC)
            expect(configParser._config.hostname).not.toContain('appium.testobject.com')
        })
    })

    describe('merge', () => {
        it('should overwrite specs if piped into cli command', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ specs: INDEX_PATH })

            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(1)
            expect(specs).toContain(INDEX_PATH)
        })

        it('should allow specifying a spec file', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec: [INDEX_PATH] })

            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(1)
            expect(specs).toContain(INDEX_PATH)
        })

        it('should allow specifying mutliple single spec file', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec : [INDEX_PATH, FIXTURES_CONF]})

            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(2)
            expect(specs).toContain(INDEX_PATH)
            expect(specs).toContain(FIXTURES_CONF)
        })

        it('should allow to specify partial matching spec file', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec : ['Plugin']})

            const specs = configParser.getSpecs()
            expect(specs).toContain(path.join(__dirname, 'initialisePlugin.test.js'))
        })

        it('should exclude duplicate spec files', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec : [INDEX_PATH, INDEX_PATH]})

            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(1)
            expect(specs).toContain(INDEX_PATH)
        })

        it('should throw if specified spec file does not exist', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            expect(() => configParser.merge({ spec: [path.resolve(__dirname, 'foobar.js')] })).toThrow()
        })

        it('should allow to specify multiple suites', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ suite: ['unit', 'functional', 'mobile']})

            const specs = configParser.getSpecs()
            expect(specs).toContain(__filename)
            expect(specs).toContain(INDEX_PATH)
            expect(specs).toContain(path.join(__dirname, 'validateConfig.test.js'))
            expect(specs).toContain(path.join(__dirname, 'detectBackend.test.js'))
        })

        it('should throw when suite is not defined', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ suite: ['blabla'] })
            expect(() => configParser.getSpecs()).toThrow(/The suite\(s\) "blabla" you specified don't exist/)
        })

        it('should throw when multiple suites are not defined', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ suite: ['blabla', 'lala'] })
            expect(() => configParser.getSpecs()).toThrow(/The suite\(s\) "blabla", "lala" you specified don't exist/)
        })

        it('should allow to specify a single suite', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ suite: ['mobile'] })

            let specs = configParser.getSpecs()
            expect(specs).toHaveLength(1)
            expect(specs).toContain(path.join(__dirname, 'detectBackend.test.js'))
        })

        it('should overwrite host and port if key are set as cli arguments', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ user: 'barfoo', key: '50fa1411-3121-4gb0-9p07-8q326vvbq7b0' })

            const config = configParser.getConfig()
            expect(config.hostname).toBe('ondemand.saucelabs.com')
            expect(config.port).toBe(443)
            expect(config.user).toBe('barfoo')
            expect(config.key).toBe('50fa1411-3121-4gb0-9p07-8q326vvbq7b0')
        })

        it('should not overwrite host and port if specified in host file', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_LOCAL_CONF)
            configParser.merge({ user: 'barfoo', key: '50fa1411-3121-4gb0-9p07-8q326vvbq7b0' })

            const config = configParser.getConfig()
            expect(config.hostname).toBe('ondemand.saucelabs.com')
            expect(config.port).toBe(443)
        })

        it('should allow specifying a exclude file', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec: [INDEX_PATH, FIXTURES_CONF] })
            configParser.merge({ exclude: [INDEX_PATH] })
            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(1)
            expect(specs).toContain(FIXTURES_CONF)
        })

        it('should allow specifying multiple exclude files', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ spec: [INDEX_PATH, FIXTURES_CONF] })
            configParser.merge({ exclude: [INDEX_PATH, FIXTURES_CONF] })
            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(0)
        })

        it('should throw if specified exclude file does not exist', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            expect(() => configParser.merge({ exclude: [path.resolve(__dirname, 'foobar.js')] })).toThrow()
        })

        it('should overwrite exclude if piped into cli command', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ exclude: [INDEX_PATH] })
            const specs = configParser.getSpecs()
            expect(specs).toHaveLength(4)
        })
    })

    describe('addService', () => {
        it('should only add functions', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.addService({
                before: undefined,
                beforeTest: 123,
                afterTest: () => 'foobar',
                after: [1, () => 'barfoo', () => 'lala']
            })

            expect(configParser._config.before).toHaveLength(0)
            expect(configParser._config.beforeTest).toHaveLength(0)
            expect(configParser._config.afterTest).toHaveLength(1)
            expect(configParser._config.after).toHaveLength(2)
        })
    })

    describe('getCapabilities', () => {
        it('allows to grab certain capabilities', () => {
            const configParser = new ConfigParser()
            configParser._capabilities = ['foo', 'bar']
            expect(configParser.getCapabilities()).toEqual(configParser._capabilities)
            expect(configParser.getCapabilities(1)).toEqual('bar')
        })
    })

    describe('getSpecs', () => {
        it('should exclude files', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const specs = configParser.getSpecs()
            expect(specs).toContain(__filename)
            expect(specs).not.toContain(path.join(__dirname, 'detectBackend.test.js'))
        })

        it('should exclude/include capability excludes', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const specs = configParser.getSpecs([INDEX_PATH], [__filename])
            expect(specs).not.toContain(__filename)
            expect(specs).not.toContain(path.join(__dirname, 'detectBackend.test.js'))
            expect(specs).toContain(INDEX_PATH)
        })

        it('should include typescript files', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const tsFile = path.resolve(FIXTURES_PATH, '*.ts')
            const specs = configParser.getSpecs([tsFile])
            expect(specs).toContain(path.resolve(FIXTURES_PATH, 'typescript.ts'))
        })

        it('should include es6 files', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const es6File = path.resolve(FIXTURES_PATH, '*.es6')
            const specs = configParser.getSpecs([es6File])
            expect(specs).toContain(path.resolve(FIXTURES_PATH, 'test.es6'))
        })

        it('should not include other file types', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const javaFile = path.resolve(FIXTURES_PATH, '*.java')
            const specs = configParser.getSpecs([javaFile])
            expect(specs).not.toContain(javaFile)
        })

        it('should include spec when specifying a suite', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)
            configParser.merge({ suite: ['mobile'], spec: [INDEX_PATH] })

            let specs = configParser.getSpecs()
            expect(specs).toHaveLength(2)
            expect(specs).toContain(INDEX_PATH)
            expect(specs).toContain(path.join(__dirname, 'detectBackend.test.js'))
        })
    })

    describe('getConfig', () => {
        it('should set proper host and port to local selenium if user and key is specified', () => {
            const configParser = new ConfigParser()
            configParser.addConfigFile(FIXTURES_CONF)

            const config = configParser.getConfig()
            expect(config.hostname).toBe('ondemand.saucelabs.com')
            expect(config.port).toBe(443)
            expect(config.user).toBe('foobar')
            expect(config.key).toBe('50fa142c-3121-4gb0-9p07-8q326vvbq7b0')
        })
    })
})
