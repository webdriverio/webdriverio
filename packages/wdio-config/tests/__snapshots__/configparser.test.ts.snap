// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigParser > disables coverage 1`] = `
{
  "autoCompileOpts": {
    "autoCompile": false,
  },
  "runner": [
    "browser",
    {
      "coverage": {
        "enabled": false,
      },
    },
  ],
}
`;

exports[`ConfigParser > disables coverage 2`] = `
{
  "autoCompileOpts": {
    "autoCompile": false,
  },
  "runner": [
    "browser",
    {
      "coverage": {
        "enabled": false,
        "statements": 100,
      },
    },
  ],
}
`;

exports[`ConfigParser > enables coverage 1`] = `
{
  "autoCompileOpts": {
    "autoCompile": false,
  },
  "runner": [
    "browser",
    {
      "coverage": {
        "enabled": true,
      },
    },
  ],
}
`;

exports[`ConfigParser > enables coverage 2`] = `
{
  "autoCompileOpts": {
    "autoCompile": false,
  },
  "runner": [
    "browser",
    {
      "coverage": {
        "enabled": true,
        "statements": 100,
      },
    },
  ],
}
`;
